### DISCUSSION TITLE
PyG installation with poetry

### DISCUSSION THREAD

**Discussion by ekosman** [pyg-team/pytorch_geometric#6815] [Q&A]
I recently moved to using poetry for managing my environment, and I was wondering what is the correct way to translate the installation.
Until now, I used conda for installing pyg via the command `conda install pyg -c pyg`. However, it seems that pyg has many dependencies that are not directly translated to poetry.

I tried to simply add `torch-geometric` to the pyproject.toml file but it doesn't seem to work well. What is your suggestion?

**Comment by rusty1s** [comment]
I would assume that poetry can also access packages from PyPi, right? In this case, `torch-geometric` should work. What is the issue you are seeing?

**Reply by ekosman** [reply]
> I would assume that poetry can also access packages from PyPi, right? In this case, `torch-geometric` should work. What is the issue you are seeing?

Indeed. However, it doesn't automatically install the related dependencies, e.g. torch-scatter, torch-sparse, pyg_lib etc.

**Comment by SauravMaheshkar** [comment]
> > I would assume that poetry can also access packages from PyPi, right? In this case, `torch-geometric` should work. What is the issue you are seeing?
> 
> Indeed. However, it doesn't automatically install the related dependencies, e.g. torch-scatter, torch-sparse, pyg_lib etc.

torch-scatter, torch-sparse, pyg_lib, .. aren't related dependencies and are usually installed separately from `torch_geometric`. For more reference, refer to the [**Installation section of the README**](https://github.com/pyg-team/pytorch_geometric#installation). You are expected to install `pyg_lib`, `torch_scatter` and `torch_sparse` before installing `torch_geometric`.

A possible solution could be adding the aforementioned repositories to your `pyproject.toml` or manually using `poetry add <>`

**Reply by rusty1s** [reply]
It is also important to note that with the upcoming release, neither of these libraries is required anymore to be installed. This hopefully unblocks this issue :)

**Reply by m-naumann** [reply]
> A possible solution could be adding the aforementioned repositories to your `pyproject.toml` or manually using `poetry add <>`

@SauravMaheshkar Unfortunately, that also does not work: https://github.com/rusty1s/pytorch_sparse/issues/156

> It is also important to note that with the upcoming release, neither of these libraries is required anymore to be installed. This hopefully unblocks this issue :)

@rusty1s I guess that won't solve the issues related to the sparse installation, or will it also?

**Reply by rusty1s** [reply]
Yeah, this would only effect PyG installation.

**Reply by ekosman** [reply]
> It is also important to note that with the upcoming release, neither of these libraries is required anymore to be installed. This hopefully unblocks this issue :)

Is there any expectation for the release date? :)

**Reply by rusty1s** [reply]
We are targeting for March 21. For now, you can also install via
```
pip install git+https://github.com/pyg-team/pytorch_geometric.git
```

**Reply by q-aaronzolnailucas** [reply]
adding `torch-scatter` and `torch-sparse` to my pyproject.toml doesn't work, I have to define a separate task to install these, since the order seems to matter: `torch` -> `torch_scatter` -> `torch_sparse` -> `torch_geometric`. Unfortunately my source code directly uses `SparseTensor`, so I couldn't avoid it. I'd welcome any working examples people have!

**Reply by rusty1s** [reply]
Yes, this is a current limitation. The only workaround right now is to install from wheels in `setup.py`, but this is not allowed by PyPi :(
